x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pipelines/nlp/tokenizers/BasicTokenizer.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.ml.inference.pipelines.nlp.tokenizers;
+
+import joptsimple.internal.Strings;
+
+import java.text.Normalizer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Basic tokenization of text by whitespace with optional extras:
+ * 1. Lower case the input
+ * 2. Convert to Unicode NFD
+ * 3. Stip accents
+ * 4. Surround CJK characters with ' '
+ *
+ * Derived from
+ * https://github.com/huggingface/transformers/blob/ba8c4d0ac04acfcdbdeaed954f698d6d5ec3e532/src/transformers/tokenization_bert.py
+ */
+public class BasicTokenizer {
```

Comment:
I mentioned this in the originating PR, but since Java strings are immutable, we are making MANY MANY different strings throughout this tokenization process. 

I do think we will either:
- use a completely different thing for tokenization
- Or build a system akin to Lucene tokenizers and analyzers. This way the underlying bytes are grabbed directly instead of copying strings all over the place. 

But, this is an optimization.

Reply:
@dimitris-athanasiou has had similar thoughts. 

This code is really easy to unit test which makes it a good candidate for optimisation. There are easy gains here.